From 60f2be9aecae52618b433045be5d71c1af98fecd Mon Sep 17 00:00:00 2001
From: nift4 <nift4@protonmail.com>
Date: Sun, 28 Aug 2022 14:02:08 +0200
Subject: [PATCH 30/34] InputManager: allow to force NULL cursor

If VNC draws cursor on client side, we do not want to have a second
cursor visible.

Change-Id: Iab1b9477a9142f7ae23484f76285981bddca1880
---
 .../android/hardware/input/IInputManager.aidl  |  1 +
 .../android/hardware/input/InputManager.java   |  9 +++++++++
 .../server/input/InputManagerService.java      | 18 ++++++++++++++++++
 3 files changed, 28 insertions(+)

diff --git a/core/java/android/hardware/input/IInputManager.aidl b/core/java/android/hardware/input/IInputManager.aidl
index e43dfa66c4ad..262a985d0abb 100644
--- a/core/java/android/hardware/input/IInputManager.aidl
+++ b/core/java/android/hardware/input/IInputManager.aidl
@@ -148,4 +148,5 @@ interface IInputManager {
 
     void registerCursorCallback(ICursorCallback callbacks);
     void unregisterCursorCallback(ICursorCallback callbacks);
+    void setForceNullCursor(boolean forceNullCursor);
 }
diff --git a/core/java/android/hardware/input/InputManager.java b/core/java/android/hardware/input/InputManager.java
index c62f20ad02f0..da678836457c 100644
--- a/core/java/android/hardware/input/InputManager.java
+++ b/core/java/android/hardware/input/InputManager.java
@@ -1158,6 +1158,15 @@ public final class InputManager {
             throw ex.rethrowFromSystemServer();
         }
     }
+    
+    /** @hide */
+    public void setForceNullCursor(boolean forceNullCursor) {
+        try {
+            mIm.setForceNullCursor(forceNullCursor);
+        } catch (RemoteException ex) {
+            throw ex.rethrowFromSystemServer();
+        }
+    }
 
     /**
      * Request or release pointer capture.
diff --git a/services/core/java/com/android/server/input/InputManagerService.java b/services/core/java/com/android/server/input/InputManagerService.java
index 98d1ad36b2f2..f20604906d57 100644
--- a/services/core/java/com/android/server/input/InputManagerService.java
+++ b/services/core/java/com/android/server/input/InputManagerService.java
@@ -270,6 +270,7 @@ public class InputManagerService extends IInputManager.Stub
     private final SparseArray<CursorCallbackRecord> mCursorCallbacks = new SparseArray<>();
     private final List<CursorCallbackRecord> mTempCursorCallbacksToNotify =
             new ArrayList<>();
+    private boolean mForceNullCursor = false;
 
     // State for the currently installed input filter.
     final Object mInputFilterLock = new Object();
@@ -501,6 +502,15 @@ public class InputManagerService extends IInputManager.Stub
         }
     }
 
+    @Override // Binder call
+    public void setForceNullCursor(boolean forceNullCursor) {
+        if (DEBUG) {
+            Slog.d(TAG, "setForceNullCursor: forceNullCursor=" + forceNullCursor + " callingPid="
+                    + Binder.getCallingPid());
+        }
+        mForceNullCursor = forceNullCursor;
+    }
+
     @Override // Binder call
     public void registerCursorCallback(ICursorCallback listener) {
         if (listener == null) {
@@ -2427,6 +2437,10 @@ public class InputManagerService extends IInputManager.Stub
     @Override
     public void setPointerIconType(int iconId) {
         deliverCursorChanged(iconId, null);
+        if (mForceNullCursor) {
+            nativeSetPointerIconType(mPtr, PointerIcon.TYPE_NULL);
+            return;
+        }
         nativeSetPointerIconType(mPtr, iconId);
     }
 
@@ -2435,6 +2449,10 @@ public class InputManagerService extends IInputManager.Stub
     public void setCustomPointerIcon(PointerIcon icon) {
         Objects.requireNonNull(icon);
         deliverCursorChanged(PointerIcon.TYPE_CUSTOM, icon);
+        if (mForceNullCursor) {
+            nativeSetPointerIconType(mPtr, PointerIcon.TYPE_NULL);
+            return;
+        }
         nativeSetCustomPointerIcon(mPtr, icon);
     }
 
-- 
2.43.0

